# Preparing for my First Exploratory Data Analysis (EDA)

Hello, everyone! This is my first Kaggle public notebook, and I'll be honest, it might be a bit all over the place. But, I'd really appreciate your thoughts, feedback, and comments on the code, content, structure, and writing style. Your insights are highly appreciated.

The purpose of this notebook is to prepare and process the CSV file from Kaggle's dataset created by CRXXOM (https://www.kaggle.com/datasets/crxxom/daily-google-news/data) to make it ready for my project. For my first project, I won't be using Python for EDA; instead, I'll be using SQL. I aim to enhance my understanding of SQL, specifically MySQL, before delving deeper into Python's libraries. I'm simply using this notebook with the Pandas library imported for the prepare and process phases because it is faster than using spreadsheets. Additionally, it is easier to document the process when I can share it publicly.

In [1]:
# Import Pandas library
import pandas as pd

In [2]:
# Load the csv file from the source to my notebook
df = pd.read_csv("/kaggle/input/daily-google-news/2023_9.csv")
# Preview of 'df' DataFrame
df

Title  \
0      Chainlink (LINK) Falters, Hedera (HBAR) Wobble...   
1      Funds punished for owning too few Nvidia share...   
2      Crude oil prices stalled as hedge funds sold: ...   
3      Grayscale's Bitcoin Win Is Still Only Half the...   
4      I'm a Home Shopping Editor, and These Are the ...   
...                                                  ...   
51297  Slovakia's Election Could Echo in Ukraine. Her...   
51298  Things to know about the Nobel Prizes - The Wa...   
51299  After brief calm, protests against killing of ...   
51300  ‘No one is safe’: France vows action as bedbug...   
51301  Slovakia election: polls open in knife-edge vo...   

                    Publisher              DateTime  \
0           Analytics Insight  2023-08-30T06:54:49Z   
1                       ZAWYA  2023-08-30T07:15:59Z   
2                       ZAWYA  2023-08-30T07:31:31Z   
3                   Bloomberg  2023-08-30T10:38:40Z   
4      Better Homes & Gardens  2023-08-30T11:00:00Z   
...                       ...                   ...   
51297      The New York Times  2023-09-30T04:01:14Z   
51298     The Washington Post  2023-09-30T04:26:44Z   
51299         Hindustan Times  2023-09-30T04:51:51Z   
51300                     CNN  2023-09-30T04:58:00Z   
51301            The Guardian  2023-09-30T05:00:00Z   

                                                    Link   Category  
0      https://news.google.com/articles/CBMibGh0dHBzO...   Business  
1      https://news.google.com/articles/CBMigwFodHRwc...   Business  
2      https://news.google.com/articles/CBMibGh0dHBzO...   Business  
3      https://news.google.com/articles/CBMib2h0dHBzO...   Business  
4      https://news.google.com/articles/CBMiPWh0dHBzO...   Business  
...                                                  ...        ...  
51297  https://news.google.com/articles/CBMiU2h0dHBzO...  Worldwide  
51298  https://news.google.com/articles/CBMimQFodHRwc...  Worldwide  
51299  https://news.google.com/articles/CBMikgFodHRwc...  Worldwide  
51300  https://news.google.com/articles/CBMiTmh0dHBzO...  Worldwide  
51301  https://news.google.com/articles/CBMieWh0dHBzO...  Worldwide  

[51302 rows x 5 columns]

It appears that the table consists of 51,302 records.

 Now, let's check the possibility of having records with null values.

In [3]:
# Checking for the possibility of columns with null values
df.isnull().sum()

Title        0
Publisher    0
DateTime     0
Link         0
Category     0
dtype: int64

There are no null values in the records.

Now, let's examine the structure of the DataFrame.

In [4]:
# Checking the DataFrame structure
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51302 entries, 0 to 51301
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Title      51302 non-null  object
 1   Publisher  51302 non-null  object
 2   DateTime   51302 non-null  object
 3   Link       51302 non-null  object
 4   Category   51302 non-null  object
dtypes: object(5)
memory usage: 2.0+ MB


Since all columns have non-null values by default, which is a good start for data cleaning. 

I'll proceed to a minor cleanup by removing the 'T' and 'Z' from the DateTime column.

In [5]:
# Convert the 'DateTime' column to datetime format
df['DateTime'] = pd.to_datetime(df['DateTime'], format='%Y-%m-%dT%H:%M:%SZ')

# Format the column (removing the 'T' and 'Z' ('YYYY-MM-DD HH:MM:SS')
df['DateTime'] = df['DateTime'].dt.strftime('%Y-%m-%d %H:%M:%S')

In [6]:
# Reviewing the condition
df

Title  \
0      Chainlink (LINK) Falters, Hedera (HBAR) Wobble...   
1      Funds punished for owning too few Nvidia share...   
2      Crude oil prices stalled as hedge funds sold: ...   
3      Grayscale's Bitcoin Win Is Still Only Half the...   
4      I'm a Home Shopping Editor, and These Are the ...   
...                                                  ...   
51297  Slovakia's Election Could Echo in Ukraine. Her...   
51298  Things to know about the Nobel Prizes - The Wa...   
51299  After brief calm, protests against killing of ...   
51300  ‘No one is safe’: France vows action as bedbug...   
51301  Slovakia election: polls open in knife-edge vo...   

                    Publisher             DateTime  \
0           Analytics Insight  2023-08-30 06:54:49   
1                       ZAWYA  2023-08-30 07:15:59   
2                       ZAWYA  2023-08-30 07:31:31   
3                   Bloomberg  2023-08-30 10:38:40   
4      Better Homes & Gardens  2023-08-30 11:00:00   
...                       ...                  ...   
51297      The New York Times  2023-09-30 04:01:14   
51298     The Washington Post  2023-09-30 04:26:44   
51299         Hindustan Times  2023-09-30 04:51:51   
51300                     CNN  2023-09-30 04:58:00   
51301            The Guardian  2023-09-30 05:00:00   

                                                    Link   Category  
0      https://news.google.com/articles/CBMibGh0dHBzO...   Business  
1      https://news.google.com/articles/CBMigwFodHRwc...   Business  
2      https://news.google.com/articles/CBMibGh0dHBzO...   Business  
3      https://news.google.com/articles/CBMib2h0dHBzO...   Business  
4      https://news.google.com/articles/CBMiPWh0dHBzO...   Business  
...                                                  ...        ...  
51297  https://news.google.com/articles/CBMiU2h0dHBzO...  Worldwide  
51298  https://news.google.com/articles/CBMimQFodHRwc...  Worldwide  
51299  https://news.google.com/articles/CBMikgFodHRwc...  Worldwide  
51300  https://news.google.com/articles/CBMiTmh0dHBzO...  Worldwide  
51301  https://news.google.com/articles/CBMieWh0dHBzO...  Worldwide  

[51302 rows x 5 columns]

With the columns cleaned and formatted to my preferences, I'll now proceed to adjust the column names and their order.

In [7]:
# Rename the 'DateTime' column to 'publishtime' to provide more context. Modify other columns to lowercase to adhere to SQL standards
df = df.rename(columns={'DateTime': 'publishtime', 'Title': 'title', 'Publisher':'publisher', 'Link':'link', 'Category':'category'})

In [8]:
# Rearrange the columns' order
df = df[['publishtime', 'publisher', 'category', 'link', 'title']]

In [9]:
# Reviewing the current condition
df

publishtime               publisher   category  \
0      2023-08-30 06:54:49       Analytics Insight   Business   
1      2023-08-30 07:15:59                   ZAWYA   Business   
2      2023-08-30 07:31:31                   ZAWYA   Business   
3      2023-08-30 10:38:40               Bloomberg   Business   
4      2023-08-30 11:00:00  Better Homes & Gardens   Business   
...                    ...                     ...        ...   
51297  2023-09-30 04:01:14      The New York Times  Worldwide   
51298  2023-09-30 04:26:44     The Washington Post  Worldwide   
51299  2023-09-30 04:51:51         Hindustan Times  Worldwide   
51300  2023-09-30 04:58:00                     CNN  Worldwide   
51301  2023-09-30 05:00:00            The Guardian  Worldwide   

                                                    link  \
0      https://news.google.com/articles/CBMibGh0dHBzO...   
1      https://news.google.com/articles/CBMigwFodHRwc...   
2      https://news.google.com/articles/CBMibGh0dHBzO...   
3      https://news.google.com/articles/CBMib2h0dHBzO...   
4      https://news.google.com/articles/CBMiPWh0dHBzO...   
...                                                  ...   
51297  https://news.google.com/articles/CBMiU2h0dHBzO...   
51298  https://news.google.com/articles/CBMimQFodHRwc...   
51299  https://news.google.com/articles/CBMikgFodHRwc...   
51300  https://news.google.com/articles/CBMiTmh0dHBzO...   
51301  https://news.google.com/articles/CBMieWh0dHBzO...   

                                                   title  
0      Chainlink (LINK) Falters, Hedera (HBAR) Wobble...  
1      Funds punished for owning too few Nvidia share...  
2      Crude oil prices stalled as hedge funds sold: ...  
3      Grayscale's Bitcoin Win Is Still Only Half the...  
4      I'm a Home Shopping Editor, and These Are the ...  
...                                                  ...  
51297  Slovakia's Election Could Echo in Ukraine. Her...  
51298  Things to know about the Nobel Prizes - The Wa...  
51299  After brief calm, protests against killing of ...  
51300  ‘No one is safe’: France vows action as bedbug...  
51301  Slovakia election: polls open in knife-edge vo...  

[51302 rows x 5 columns]

Cool. I like this format because it is met the SQL standard

In [10]:
# Take a look about the posibility of duplicates record
duplicate_count = df.duplicated()

duplicate_count.sum()

4206

In [11]:
# Remove the duplicate from the DataFrame
# Remove duplicate rows
df = df.drop_duplicates()

# Reset the index after removing duplicates
df = df.reset_index(drop=True)

In [12]:
# Reviewing our DataFrame again
df

publishtime               publisher   category  \
0      2023-08-30 06:54:49       Analytics Insight   Business   
1      2023-08-30 07:15:59                   ZAWYA   Business   
2      2023-08-30 07:31:31                   ZAWYA   Business   
3      2023-08-30 10:38:40               Bloomberg   Business   
4      2023-08-30 11:00:00  Better Homes & Gardens   Business   
...                    ...                     ...        ...   
47091  2023-09-30 04:01:14      The New York Times  Worldwide   
47092  2023-09-30 04:26:44     The Washington Post  Worldwide   
47093  2023-09-30 04:51:51         Hindustan Times  Worldwide   
47094  2023-09-30 04:58:00                     CNN  Worldwide   
47095  2023-09-30 05:00:00            The Guardian  Worldwide   

                                                    link  \
0      https://news.google.com/articles/CBMibGh0dHBzO...   
1      https://news.google.com/articles/CBMigwFodHRwc...   
2      https://news.google.com/articles/CBMibGh0dHBzO...   
3      https://news.google.com/articles/CBMib2h0dHBzO...   
4      https://news.google.com/articles/CBMiPWh0dHBzO...   
...                                                  ...   
47091  https://news.google.com/articles/CBMiU2h0dHBzO...   
47092  https://news.google.com/articles/CBMimQFodHRwc...   
47093  https://news.google.com/articles/CBMikgFodHRwc...   
47094  https://news.google.com/articles/CBMiTmh0dHBzO...   
47095  https://news.google.com/articles/CBMieWh0dHBzO...   

                                                   title  
0      Chainlink (LINK) Falters, Hedera (HBAR) Wobble...  
1      Funds punished for owning too few Nvidia share...  
2      Crude oil prices stalled as hedge funds sold: ...  
3      Grayscale's Bitcoin Win Is Still Only Half the...  
4      I'm a Home Shopping Editor, and These Are the ...  
...                                                  ...  
47091  Slovakia's Election Could Echo in Ukraine. Her...  
47092  Things to know about the Nobel Prizes - The Wa...  
47093  After brief calm, protests against killing of ...  
47094  ‘No one is safe’: France vows action as bedbug...  
47095  Slovakia election: polls open in knife-edge vo...  

[47096 rows x 5 columns]

Looks like the record of new DataFrame is matched with the manual calculation.
51302 rows - 4206 rows = 47096 rows

Lets short the data by 'publishdate' column

In [13]:
# Sort the DataFrame by 'publishtime' in descending order (from the newest to the oldest)
df = df.sort_values(by='publishtime', ascending=False)

# Select the top 10 newest records
newest_records = df.head(10)

# Reviewing newest records
newest_records

publishtime            publisher   category  \
34839  2023-09-30 05:26:42              MLB.com     Sports   
19838  2023-09-30 05:15:02                 WION  Headlines   
27319  2023-09-30 05:11:35                 WION    Science   
19837  2023-09-30 05:09:00  The Washington Post  Headlines   
19836  2023-09-30 05:01:46  The Washington Post  Headlines   
47095  2023-09-30 05:00:00         The Guardian  Worldwide   
47094  2023-09-30 04:58:00                  CNN  Worldwide   
34838  2023-09-30 04:53:07           CBS Sports     Sports   
34837  2023-09-30 04:52:30         Golf Monthly     Sports   
47093  2023-09-30 04:51:51      Hindustan Times  Worldwide   

                                                    link  \
34839  https://news.google.com/articles/CBMiQmh0dHBzO...   
19838  https://news.google.com/articles/CCAiC082a0UzZ...   
27319  https://news.google.com/articles/CBMiZ2h0dHBzO...   
19837  https://news.google.com/articles/CBMiYGh0dHBzO...   
19836  https://news.google.com/articles/CBMiT2h0dHBzO...   
47095  https://news.google.com/articles/CBMieWh0dHBzO...   
47094  https://news.google.com/articles/CBMiTmh0dHBzO...   
34838  https://news.google.com/articles/CBMilgFodHRwc...   
34837  https://news.google.com/articles/CBMikAFodHRwc...   
47093  https://news.google.com/articles/CBMikgFodHRwc...   

                                                   title  
34839   Postseason eludes Padres despite late-season run  
19838  US: House Republicans reject own bill to exten...  
27319  Astronomers capture first direct evidence of B...  
19837  Prosecutors cite Trump's supposed gun purchase...  
19836  House fails to pass short-term funding bill as...  
47095  Slovakia election: polls open in knife-edge vo...  
47094  ‘No one is safe’: France vows action as bedbug...  
34838  Utah vs. Oregon State score: No. 10 Utes lifel...  
34837  Ryder Cup Day 2 Latest Scores - McIlroy And Fl...  
47093  After brief calm, protests against killing of ...

In [14]:
# Select the bottom 10 oldest records
oldest_records = df.tail(10)

# Reviewing oldest records
oldest_records

publishtime             publisher       category  \
23893  2023-08-29 10:05:40         WRAL TechWire        Science   
23892  2023-08-29 09:30:01              Phys.org        Science   
23891  2023-08-29 09:20:37         New Scientist        Science   
23890  2023-08-29 01:00:09              Upworthy        Science   
23889  2023-08-28 17:23:35  Kennedy Space Center        Science   
23888  2023-08-25 16:21:27  Kennedy Space Center        Science   
34840  2023-08-25 13:01:33         IEEE Spectrum     Technology   
41073  2022-11-10 07:10:30              Sky News      Worldwide   
6337   2019-11-10 05:15:43          KarmaWeather  Entertainment   
41072  2019-03-10 20:53:31    The New York Times      Worldwide   

                                                    link  \
23893  https://news.google.com/articles/CBMifWh0dHBzO...   
23892  https://news.google.com/articles/CBMiVmh0dHBzO...   
23891  https://news.google.com/articles/CBMia2h0dHBzO...   
23890  https://news.google.com/articles/CBMiW2h0dHBzO...   
23889  https://news.google.com/articles/CBMifGh0dHBzO...   
23888  https://news.google.com/articles/CBMia2h0dHBzO...   
34840  https://news.google.com/articles/CBMiJmh0dHBzO...   
41073  https://news.google.com/articles/CBMiVWh0dHBzO...   
6337   https://news.google.com/articles/CBMiLGh0dHBzO...   
41072  https://news.google.com/articles/CBMiIWh0dHBzO...   

                                                   title  
23893  Duke engineers unlock quantum secrets in molec...  
23892  Quantum-enhanced detection for chip-scale wire...  
23891  Ancient humans may have worn shoes more than 1...  
23890  How 'light pollution' has turned looking at a ...  
23889                         SpaceX Falcon Heavy Pysche  
23888                              OSIRIS-REx Homecoming  
34840  Meta’s Flamera Has a New Vision for Augmented ...  
41073  Ukraine war latest: Poland summons ambassador ...  
6337                     Today's Daily Chinese Horoscope  
41072  A Tragic Fire and Broken Promises in South Africa

Looks like the records is containt news from 2023-08-25 13:01:33 to 2023-09-30 05:26:42. The 3 records outside the range is strange (Index number: 41073, 6337, 41072). I assume it is because the publisher re-used their old pages to write the new news. Because we cant edit those three into the correct publishdate we need to drop them from our DataFrame.

To make our DataFrame even more balanced, let's define a one-month period. 
While over 90% of the records are from September news, the latest record from September is on '2023-09-30 05:26:42'. 

Let's pretend we're working for a popular news agency, and the stakeholders want a report on news data, including news from all competitors, for the last 30 days. However, they want the report on the last day of the month. To meet this requirement, we will exclude the data from 2023-09-30. To ensure fairness and consistency in future reporting, we'll also include records from the last day of the previous month, which is 2023-08-31. This way, in the next month, our reporting will be consistent and equitable.

With that in mind, our chosen time frame is from '2023-08-31 00:00:00' to '2023-09-29 23:59:59'.

Let's remove all records outside of this time frame!

In [15]:
# Convert 'publishtime' to datetime format
df['publishtime'] = pd.to_datetime(df['publishtime'])

# Create a boolean mask named sept_records for the date range
start_date = pd.to_datetime('2023-08-31 00:00:00')
end_date = pd.to_datetime('2023-09-29 23:59:59')
sept_records = (df['publishtime'] >= start_date) & (df['publishtime'] <= end_date)

# Use the mask(sept_records) to select the records within the specified date range
df= df[sept_records]

# Reviewing the current DataFrame
df

publishtime            publisher       category  \
19793 2023-09-29 23:58:11     CBS Evening News      Headlines   
41064 2023-09-29 23:53:42         GLYFE Nation     Technology   
47072 2023-09-29 23:53:00         The Guardian      Worldwide   
19792 2023-09-29 23:53:00         The Guardian      Headlines   
6313  2023-09-29 23:52:34      CNBC Television       Business   
...                   ...                  ...            ...   
6360  2023-08-31 00:08:00  Yahoo Entertainment  Entertainment   
19916 2023-08-31 00:05:59   Medical News Today         Health   
13000 2023-08-31 00:04:00       New York Post       Headlines   
18    2023-08-31 00:00:35      Financial Times       Business   
34872 2023-08-31 00:00:00            Pure Xbox     Technology   

                                                    link  \
19793  https://news.google.com/articles/CCAiC0FBeUQ0V...   
41064  https://news.google.com/articles/CCAiC3o4ZENmO...   
47072  https://news.google.com/articles/CBMiUWh0dHBzO...   
19792  https://news.google.com/articles/CBMiUWh0dHBzO...   
6313   https://news.google.com/articles/CCAiCzdWazQ2e...   
...                                                  ...   
6360   https://news.google.com/articles/CBMiVGh0dHBzO...   
19916  https://news.google.com/articles/CBMia2h0dHBzO...   
13000  https://news.google.com/articles/CBMiYWh0dHBzO...   
18     https://news.google.com/articles/CBMiP2h0dHBzO...   
34872  https://news.google.com/articles/CBMiYmh0dHBzO...   

                                                   title  
19793  Record-breaking rainfall causes flooding in Ne...  
41064  Disney Speedstorm Season 4: Unveiling A Whole ...  
47072  White House warns of ‘unprecedented’ Serbian t...  
19792  White House warns of ‘unprecedented’ Serbian t...  
6313   Don't trade stocks on what you're expecting fr...  
...                                                  ...  
6360   Dukes of Hazzard star John Schneider reveals '...  
19916  Atrial fibrillation: Stress, insomnia linked t...  
13000  DOJ contacted Hunter Biden, whistleblower team...  
18     Will a rare strike threaten the 'buy Japan' mo...  
34872  Starfield Director Todd Howard Praises Xbox In...  

[46606 rows x 5 columns]

Looks great. I think we're almost done. Let's save the current DataFrame into new CSV and JSON files for EDA in MySQL.

In [16]:
# Use 'gn_sept' (Google News - September) as the new name for my CSV file
new_name = 'gn_sept.csv'

# Use the 'to_csv' method to save 'df' DataFrame to a CSV file
df.to_csv(new_name, index=False)  # Set index=False to exclude the index column

In [17]:
# Use 'gn_sept' (Google News - September) as the new name for my JSON file
new_name2 = 'gn_sept.json'

# Use the 'to_json' method to save the DataFrame to a JSON file
df.to_json(new_name2, orient='records')

Seems like my MySQL Workbench is too laggy to import CSV or JSON files, so I will create a new text file. Then, I will manually copy the contents from the file and input them using queries in MySQL Workbench.

In [18]:
# Create an empty list 'y' to store the converted rows as tuples
y = []

# Loop through the rows of 'df' DataFrame
for i in range(len(df)):
    # Convert the row into a tuple
    x = tuple(df.iloc[i])
    # Append the tuple to the 'y' list
    y.append(x)


# Uncomment '#y' to view its contents
#y

In [19]:
# Create a text file 'gn_sept.txt' for writing
file = open('gn_sept.txt', 'w')

# Loop through the list of tuples and write each tuple to the text file
for tuple in y:
    file.write(str(tuple) + ',' + '\n')

# Close the text file to save the changes
file.close()

Because I won't be using Python (Pandas) for my first EDA, this concludes my notebook. If you want to practice your EDA skills with this notebook, feel free to use it.





Goodbye for now, and thank you for reading.

Note: I wrote this notebook before the October data update, so I didn't include the data for October.